portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js

```diff
@@ -1,161 +1,52 @@
- * @returns {Promise<object>} - azure storage details with folder & file name, full path & url.
- */
-const saveFileToAzure = async (file, bankId) => {
-  try {
-    console.info(`Attempting to save utilisation report to Azure for bank: ${bankId}`);
-    const { originalname, buffer } = file;
-
-    const fileInfo = await uploadFile({
-      fileshare: FILESHARES.UTILISATION_REPORTS,
-      folder: bankId,
-      filename: formatFilenameForSharepoint(originalname),
-      buffer,
-      allowOverwrite: true,
-    });
-
-    if (!fileInfo || fileInfo.error) {
-      throw new Error(`Failed to save utilisation report to Azure - ${fileInfo?.error?.message ?? 'cause unknown'}`);
-    }
-
-    console.info(`Successfully saved utilisation report to Azure for bank: ${bankId}`);
-    return fileInfo;
-  } catch (error) {
-    console.error('Failed to save utilisation report to Azure ', error);
-    throw error;
-  }
-};
+const { InvalidReportStatusError } = require('../../errors');
 
 const uploadReportAndSendNotification = async (req, res) => {
   try {
     const { file } = req;
 
-    const { formattedReportPeriod, reportData, reportPeriod, user } = req.body;
-    const parsedReportData = JSON.parse(reportData);
-    const parsedUser = JSON.parse(user);
-    const parsedReportPeriod = JSON.parse(reportPeriod);
-
     if (!file) {
       return res.status(400).send();
     }
 
+    const { formattedReportPeriod, reportData, reportPeriod, user } = req.body;
+    const parsedReportData = JSON.parse(reportData);
+    const parsedUser = JSON.parse(user);
+    const parsedReportPeriod = JSON.parse(reportPeriod);
     const bankId = parsedUser?.bank?.id;
 
-    const existingReports = await api.getUtilisationReports(bankId, {
-      reportPeriod: parsedReportPeriod,
-    });
-
-    if (existingReports.length !== 1) {
-      return res.status(500).send(`Expected 1 report but found ${existingReports.length} with bank ID ${bankId} and report period '${parsedReportPeriod}'`);
-    }
-
-    const existingReport = existingReports[0];
+    const reportId = await validateReportForPeriodIsInReportNotReceivedStateAndReturnId(bankId, parsedReportPeriod);
+    const azureFileInfo = await saveUtilisationReportFileToAzure(file, bankId);
+    const { dateUploaded } = await api.saveUtilisationReport(reportId, parsedReportData, parsedUser, azureFileInfo);
 
-    if (existingReport.status !== UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED) {
-      return res
-        .status(500)
-        .send(
-          `Expected report to be in '${UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED}' state (was actually in '${existingReport.status}' state)`,
-        );
+    try {
+      await sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam(parsedUser?.bank?.name, formattedReportPeriod);
+    } catch(error) {
+      console.error('Failed to send report upload notification to ukef gef reporting team %o', error);
```

Comment:
This catching here is a change to functionality, it just felt weird to be showing the user an error page when this emails fails to send given they are not told it was sending it in the first place and otherwise the upload went fine. But if we don't want this I can change this back. I thought about changing the other email to also fail quietly from the user's perspective but left it since we display the email addresses on the page so that would have required more thinking.

Reply:
Personally, I think the failing silently is something we want to add to the function being called rather than in the controller (a try-catch inside the try-catch feels a bit strange to read). My preferred method would be to have two methods exported from the emails service - one which can throw and has the name `sendUtilisation...OrThrow` and one which doesn't, but I'm not sure if that's the best solution here

Reply:
Originally I had it in the downstream method but the issue I felt was then it doesn't match the behaviour of the other send email method, and also both send multiple emails and could throw sending one and not the others as well so 'OrThrow' seems a little misleading in and of itself if I add it to the other method...